CHANGELOG.md

```diff
@@ -15,6 +15,14 @@
 - Added a [guide](docs/devctr-image.md) for updating the dev container image.
 - Added a new API call, `PUT /mmds/config`, for configuring the
   `MMDS` with a custom valid link-local IPv4 address.
+- Added `track_dirty_pages` field to `machine-config`, which specifies
+  whether dirty pages should be tracked and which allows incremental
+  snapshots to be created.
+- Added a new API call, `PATCH /vm`, for changing the microVM state (to
+  `Paused` or `Resumed`).
+- Added a new API call, `PUT /snapshot/create`, for creating a full or diff
```

Comment:
The [design proposal](https://github.com/firecracker-microvm/firecracker/blob/9d427b33d989c3225d874210f6c2849465941dc0/docs/snapshotting/design.md) introduces separate APIs for creating incremental and full snapshots. Are you proposing a single endpoint ?

Reply:
Yes. Because both `/create` and `/create-diff` would have the same request body, I was thinking to make that distinction via another field in the body and reduce this way the swagger file size (and probably avoid also some duplicate code).

Reply:
One of the reasons they are separated was to provide a clear distinction eliminating the need to have a parameter for that. This is also future looking, for example if in the future you would add another parameter that only makes sense for `create` it would create some extra complexity by requiring extra conditionals as well as json ser/de issues for `create-diff` requests.

Reply:
Are you suggesting that, when `create` and `create-diff` will diverge (from the request bodies point of view) we will add a new `SnapshotCreateConfig`, `SnapshotCreateConfigDiff` for example, or that we should make this distinction at this moment, i.e. have 2 different structs with the same fields? From what I understand from the design proposal RFC, these requests have the same body, so I supposed they will also use the same struct for ser/de.

Reply:
The main concern here is extensibility and complexity. From my point of view things are much more simpler whit as few parameters to configure as possible.

Reply:
What I was trying to say is that when those requests will have different bodies, we will also have to add some new `SnapshotCreateConfig` struct besides the existing one (or more precisely, we will rename the existing one to something like `SnapshotCreateConfigDiff` and create a new one with all the fields required for the updated `create` action). This also means some extra code. I am okay with having 2 different APIs for creating incremental and full snapshots, I just wanted to make sure that at this moment, in this PR, the struct that will be used for ser/de for both requests should be the same and not have a duplicate (`SnapshotCreateConfig` and `SnapshotCreateConfigDiff` - both with the same fields).

Reply:
> I just wanted to make sure that at this moment, in this PR, the struct that will be used for ser/de for both requests should be the same and not have a duplicate (`SnapshotCreateConfig` and `SnapshotCreateConfigDiff` - both with the same fields).

Definetly only one: SnapshotCreateConfig.

Reply:
I think a diff snapshot is a special case of a snapshot. Since the `load` API only takes full snapshots, does it make sense to have the default of `snapshot/create` create a full snapshot as well? This way the two operations are by default mirrored in a sense.

I have a few comments on naming as well.
There is no need to name them `..Config`. Besides Mmds (which is named wrong, ooopsi), all the other objects are named without that sufix. I see no reason to keep the suffix for snapshot.
Nevertheless, naming will be a bit awkward for the resource because `Snapshot` could be seen as a resource, while `snapshot/create` and `snapshot/load` cannot. To make it less awkward from the REST API point of view, can we use the same body for both requests instead? I would go as far as having the Snapshot resource (on `/snapshot`) and then an `action` field enum with the values load and create. We can make sure in the code that for load we ignore everything else that is not the memory path and the state path.

```swagger
Snapshot:
  type: object
  required:
    - snapshot_path
    - mem_file_path
    - action
  properties:
    version:
      type: integer
      minimum: TBD
      maximum: TBD
      description: Optional field for snapshot format version. If omitted
                   Firecracker will default to creating a snapshot with the
                   current version.
    snapshot_path:
      type: string
      description: Path to file which will be populated with the vmm state.
    mem_file_path:
      type: string
      description: Path to file which will be populated with the guest memory.
                   The file will contain a copy of the guest memory or copies
                   of just the dirtied pages, depending on which URL this
                   object is used.
    type: string
         enum: [full, diff]
         description: Type of snapshot to create. By default a full snapshot is created. This field
          only makes sense for the create operation and is ignored otherwise.
    action: string
        enum: [load, create]
        description: Action to take on the snapshot resource.
    support_inc_snapshots: bool
        description: Support incremental snapshots. This only makes sense for load operations and is ignored otherwise.
```

To create a snapshot:

1. Full snapshot:  
PUT /snapshot {action: "create", snapshot_file: "/blah/1231", "mem_file_path": "/blah/1"}

2. Diff snapshot:
PUT /snapshot {action: "create", type: "diff", snapshot_file: "/blah/1231", "mem_file_path": "/blah/1"}

3. Snapshot at a certain version:
PUT /snapshot {action: "create", version: 12, snapshot_file: "/blah/1231", "mem_file_path": "/blah/1"}

To load a snapshot:
PUT /snapshot {action: "load", snapshot_file: "/blah/1231", "mem_file_path": "/blah/1"}

I don't know if having actions is a common problem when designing restful APIs, but there are definitely other people that struggled with the same problem as we have now :)). Having the `action` field is not uncommon.

Whatever path forward we take here (either action in path or action in body), we should find some better names for `SnapshotCreate` and `SnapshotLoad` as these are rather operations than nouns. These names are going to be used by everyone using this functionality (and generating clients from our swagger), so we should try to make them less confusing.

Reply:
Idea no 2 which should be less intrusive is to rename `SnapshotCreate` and `SnapshotLoad` to `CreateSnapshotParams` and `LoadSnapshotParams`.

Reply:
Updated with `CreateSnapshotParams` and `LoadSnapshotParams`.